plugin/fulltext/ngram_parser/plugin_ngram.cc

```diff
@@ -0,0 +1,238 @@
+/*
+   Copyright (c) 2014, 2017, Oracle and/or its affiliates. All rights reserved.
+   Copyright (c) 2020, MariaDB.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License, version 2.0,
+   as published by the Free Software Foundation.
+
+   This program is also distributed with certain software (including
+   but not limited to OpenSSL) that is licensed under separate terms,
+   as designated in a particular file or component or in included license
+   documentation.  The authors of MySQL hereby grant you an additional
+   permission to link the program and your derivative works with the
+   separately licensed software that they have included with MySQL.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License, version 2.0, for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
+
+#include <stddef.h>
+#include <limits.h>
+#include <my_attribute.h>
+#include <my_global.h>
+#include "../storage/innobase/include/fts0tokenize.h"
+
+#define is_alnum(ct)    ((ct) & (_MY_U | _MY_L | _MY_NMR))
+
+static int ngram_token_size;
+
+static MYSQL_SYSVAR_INT(token_size, ngram_token_size,
+  PLUGIN_VAR_READONLY,
+  "Ngram full text plugin parser token size in characters",
+  nullptr, nullptr, 2, 1, 10, 0);
+
+// Splits a string into ngrams and emits them.
+static int split_into_ngrams(MYSQL_FTPARSER_PARAM *param, const uchar *doc,
+                             int len, MYSQL_FTPARSER_BOOLEAN_INFO *bool_info)
+{
+  const CHARSET_INFO *cs= param->cs;
+  const uchar *start= doc;
+  const uchar *end= doc + len;
+  const uchar *next= start;
+  int n_chars= 0;
+  int ret= 0;
+  bool is_first= true;
+
+  assert(cs->mbminlen == 1);
+
+  while (next < end) {
+    int char_type;
+    int char_len= my_ci_ctype(cs, &char_type, next, end);
+
+    // Broken data?
+    if (next + char_len > end || char_len == 0)
+      break;
+
+    // Avoid creating partial n-grams.
+    const bool is_underscore= (char_len == 1 && *next == '_');
+    if (!is_alnum(char_type) && !is_underscore) {
```

Comment:
Why are we computing `is_underscore` even if `!is_alnum()` holds? Why do we hard-code that code point? Originally, that code point 0x5f was ← (left arrow), and in EBCDIC, the underscore would be 0x6d.

Why is the underscore special? If it really needs to be special, should it not be part of the collation definitions?

Reply:
Underscore is often used in identifiers in programming languages as a replacement for a space, so it may be beneficial to treat it as a word character. There is already code in the codebase that does like that. For example, in `/storage/maria/ma_ftdefs.h`:
```c
#define true_word_char(ctype, character) \
                      ((ctype) & (_MY_U | _MY_L | _MY_NMR) || \
                       (character) == '_')
```
which is then used in `maria_ft_get_word()`.
That's why I tried to mimic that behavior.

In the current version code is changed to use `cs->cset->mb_wc()` to check for the particular code point. That makes it work with all supported encodings, not only with those that are ASCII-compatible.